File: src/main/scala/com/vc/hard/L57.scala
package com.vc.hard

object L57 {
  def insert(arr: Array[Array[Int]], newInterval: Array[Int]): Array[Array[Int]] = {
    var i = 0
    val n = arr.length
    import scala.collection.mutable
    val newArr = new mutable.ArrayBuffer[Array[Int]]()
    val start = newInterval(0)
    val end = newInterval(1)

    var mergeStart = start
    while(i < n && start > arr(i)(1)) {
      newArr += arr(i)
      i += 1
    }
    mergeStart = if(i < n) Math.min(start, arr(i)(0)) else start

    var mergeEnd = end
    while(i < n && end >= arr(i)(0)) {
      i += 1
    }
    mergeEnd = if(i - 1 >= 0) Math.max(end, arr(i - 1)(1)) else end

    newArr += Array(mergeStart, mergeEnd)
    while(i < n) {
      newArr += arr(i)
      i += 1
    }
    newArr.toArray
  }
}

File: src/main/java/com/vc/hard/L57.java
package com.vc.hard;

import java.util.*;

class L57 {
    public int[][] insert(int[][] intervals, int[] newInterval) {
        List<int[]> output = new ArrayList<int[]>();
        int n = intervals.length;

        int i = 0;
        int start = newInterval[0];
        int end = newInterval[1];
        while(i < n && start > intervals[i][1]) {
            output.add(intervals[i++]);
        }
        int newStart = i < n ? Math.min(start, intervals[i][0]) : start;

        while(i < n && end >= intervals[i][0]) i++;
        int newEnd = (i - 1) >= 0 ? Math.max(end, intervals[i - 1][1]) : end;

        output.add(new int[]{newStart, newEnd});
        while(i < n) {
            output.add(intervals[i++]);
        }

        int[][] res = new int[output.size()][2];
        for(int k = 0; k < output.size(); k++) {
            int[] r = output.get(k);
            res[k][0] = r[0];
            res[k][1] = r[1];
        }
        return res;
    }
}
